core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -1013,16 +1013,17 @@ object IO {
   final def bracket0[E, A, B](
     acquire: IO[E, A]
   )(release: (A, ExitResult[E, B]) => IO[Nothing, Unit])(use: A => IO[E, B]): IO[E, B] =
-    Ref[Option[(A, ExitResult[E, B])]](None).flatMap { m =>
+    Ref[Option[(A, Fiber[E, B])]](None).flatMap { m =>
       (for {
-        r <- (for {
-              a <- acquire
-              f <- use(a).fork
-              r <- f.observe
-              _ <- m.set(Some((a, r)))
-            } yield r).uninterruptibly
-        b <- r.fold(IO.now, IO.fail0, IO.terminate0)
-      } yield b).ensuring(m.get.flatMap(_.fold(unit) { case ((a, r)) => release(a, r) }))
+        f <- acquire.flatMap(a => use(a).fork.flatMap(f => m.set(Some(a -> f)).const(f))).uninterruptibly
+        b <- f.join
+      } yield b).ensuring(m.get.flatMap(_.fold(IO.unit) {
+        case (a, f) =>
+          f.tryObserve.flatMap {
+            case Some(r) => release(a, r)
+            case None    => f.interrupt
+          }
```

Comment:
> At least the more common version of bracket is going to be much faster

I still don't have a good intuition of the performance impact of using one version vs another, I mean:

- a fiber is just a few hundred bytes, forking must be super fast
- joining has virtually no cost since it merely adds a callback to a list
- `tryObserve` (should we call it `poll`? like `Promise.poll`) is super fast

If my reasoning is correct, that leaves interruption, which may block the thread indeed.